sorc/ncep_post.fd/UPP_PHYSICS.f

```diff
@@ -4501,6 +4504,473 @@ END SUBROUTINE CALSLR_UUTAH
+        endif
+      END DO                               ! end of J loop
+
+!     deallocate (wrk1, wrk2, wrk3, cosl)
+! GFS use lon avg as one scaler value for pole point
+!
+      ! call poleavg(IM,JM,JSTA,JEND,SMALL,COSL(1,jsta),SPVAL,ABSV(1,jsta))
+!
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call fullpole(absv(ista_2l:iend_2u,jsta_2l:jend_2u),avpoles)     
+!
+      cosltemp=spval
+      if(jsta== 1) cosltemp(1:im, 1)=coslpoles(1:im,1)
+      if(jend==jm) cosltemp(1:im,jm)=coslpoles(1:im,2)
+      avtemp=spval
+      if(jsta== 1) avtemp(1:im, 1)=avpoles(1:im,1)
+      if(jend==jm) avtemp(1:im,jm)=avpoles(1:im,2)
+!        
+      call poleavg(IM,JM,JSTA,JEND,SMALL,cosltemp(1,jsta),SPVAL,avtemp(1,jsta))
+!
+      if(jsta== 1) absv(ista:iend, 1)=avtemp(ista:iend, 1)
+      if(jend==jm) absv(ista:iend,jm)=avtemp(ista:iend,jm)
+!    
+!        deallocate (wrk1, wrk11, wrk2, wrk3, cosl, iw, ie)
+!    
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call exch(div(ista_2l:iend_2u,jsta_2l:jend_2u))
+!
+! store absv and div factors before poisson loops
+!$omp parallel do private(i,j)
+      do j=jsta,jend
+      do i=ista,iend
+        atmp(i,j)=0.25*(absv(i,j)-f(i,j))/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*cosl(i,j)*4.)
+        dtmp(i,j)=0.25*div(i,j)/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*cosl(i,j)*4.)
+      enddo
+      enddo
+!
+! poisson solver for psi and chi 
+      psi=0.
+      ta=mpi_wtime()
+      do jjk=1,1000
+      do jj=1,300 
+        call exch(psi(ista_2l:iend_2u,jsta_2l:jend_2u))
+        ptmp=psi
+        do j=jsta,jend
+        do i=ista,iend
+          if (j>1 .and. j<jm) then
+            psi(i,j) = 0.25*(ptmp(i-1,j)+ptmp(i+1,j)+ptmp(i,j-1)+ptmp(i,j+1))-atmp(i,j)
+            edif=psi(i,j)-pval
+            edif=abs(edif)
+            err=max(edif,err)
+          endif
+        enddo
+        enddo
+        if (jsta==1) then
+          psi(iend,1)=0.
+          do i=ista,iend-1
+            psi(iend,1)=psi(iend,1)+psi(i,2)
+          enddo
+          do i=ista,iend
+            psi(i,1)=psi(iend,1)/(iend-ista)
+          enddo
+        endif
+        if (jend==jm) then
+          psi(iend,jm)=0.
+          do i=ista,iend-1
+            psi(iend,jm)=psi(iend,jm)+psi(i,jm-1)
+          enddo
+          do i=ista,iend
+            psi(i,jm)=psi(iend,jm)/(iend-ista)
+          enddo
+        endif
+      enddo   ! end of jj loop for psi
+      call mpi_allreduce (err,errmax,1,mpi_real,mpi_max,mpi_comm_comp,ier)
+        if(  errmax .lt. 50.)  then
+            exit
+            endif
+      enddo    ! end of jjk loop for psi
+!
+      chi=0.
+      tb=mpi_wtime()
+      do jjk=1,1000
+      do jj=1,300 
+        call exch(chi(ista_2l:iend_2u,jsta_2l:jend_2u))
+        ptmp=chi
+        do j=jsta,jend
+        do i=ista,iend
+          if (j>1 .and. j<jm) then
+            chi(i,j) = 0.25*(ptmp(i-1,j)+ptmp(i+1,j)+ptmp(i,j-1)+ptmp(i,j+1))-dtmp(i,j)
+            edif=psi(i,j)-pval
```

Comment:
@KarinaAsmar-NOAA @JesseMeng-NOAA What is the 'pval' used for? I don't see it initialized or calculated?

Reply:
should be
edit=chi(i,j)-ptmp(i,j)